File: 0323-number-of-connected-components-in-an-undirected-graph/dfs_approach.py
# Time complexity: O(V + E), Space complexity: O(V) for visited, O(E) for building  graph + Auxiliary space : O(V) = O(V + E)
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        num_comps = 0
        graph = [[] for _ in range(n)]
        visited = [False]*n

        for e in edges:
            graph[e[0]].append(e[1])
            graph[e[1]].append(e[0])

        def dfs(node):
            visited[node] = True

            for neigh in graph[node]:
                if not visited[neigh]:
                    dfs(neigh)
        
        for i in range(n):
            if not visited[i]:
                num_comps += 1
                dfs(i)

        return num_comps

File: codingninjas_dfs_traversal/dfs_approach.cpp
// Time complexity: O(V + E), Space complexity: O(V + E)
vector<int> compDFS(int node, vector<bool>& visited, vector<int> graph[], vector<int>& trav) {
    visited[node] = true;
    trav.push_back(node);

    for (int neigh : graph[node]) {
        if (!visited[neigh]) {
            trav = compDFS(neigh, visited, graph, trav);
        }
    }

    return trav;
}

vector<vector<int>> depthFirstSearch(int V, int E, vector<vector<int>> &edges)
{
    vector<vector<int>> dfsTrav;
    vector<bool> visited(V);
    vector<int> graph[V];

    for (auto edge : edges) {
        int u = edge[0], v = edge[1];
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    for (int i = 0; i < V; i++) {
        if (!visited[i]) {
            vector<int> trav;
            trav = compDFS(i, visited, graph, trav);
            dfsTrav.push_back(trav);
        }
    }


    return dfsTrav;
}